app/services/doc_auth/error_generator.rb

```diff
@@ -189,14 +198,43 @@ def get_error_messages(liveness_enabled, response_info)
         end
       end
 
-      portrait_match_results = response_info[:portrait_match_results] || {}
-      if liveness_enabled && portrait_match_results.dig(:FaceMatchResult) != 'Pass'
-        errors[SELFIE] << Errors::SELFIE_FAILURE
+      selfie_error = get_selfie_error(liveness_enabled, response_info)
+      if liveness_enabled && !!selfie_error
+        errors[SELFIE] << selfie_error
       end
 
       errors
     end
 
+    def get_selfie_error(liveness_enabled, response_info)
+      # The part of the response that contains information about the selfie
+      portrait_match_results = response_info[:portrait_match_results] || {}
+      # The overall result of the selfie, 'Pass' or 'Fail'
+      face_match_result = portrait_match_results.dig(:FaceMatchResult)
+      # The reason for failure (if it failed), also sometimes contains success info
+      face_match_error = portrait_match_results.dig(:FaceErrorMessage)
+
+      # No error if liveness is not enabled or if there's no failure
+      if !liveness_enabled || !face_match_result || face_match_result == 'Pass'
+        return nil
+      end
+
+      case face_match_error
+      # Error when the image on the id does not match the selfie image, but the image was acceptable
+      when 'Successful. Liveness: Live'
+        return Errors::SELFIE_FAILURE
```

Comment:
This is the same check as in the new selfie concern, but I didn't see a great way to share the code.

Reply:
include the concern here?

Reply:
I thought of that, but decided it was going to feel too weird. I'll give that another shot, thanks for the nudge.

Reply:
That was a good idea, thanks! Did that [here](https://github.com/18F/identity-idp/pull/9941/commits/b86e5f89d2d371c48673318dd7447ba5531c0c27).